File: PROFIBUS-Project/profibus-npm/src/gsdParser.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { XMLParser } from "fast-xml-parser";

export type GSDEntry = {
  filepath: string;
  metadata: Record<string, any>;
  io_config: Record<string, any>;
};

const parser = new XMLParser({
  ignoreAttributes: false,
  attributeNamePrefix: "",
});

export function loadGSD(filepath: string, basePath?: string): GSDEntry {
  let file = filepath;
  if (basePath && !file.startsWith("/")) {
    file = resolve(basePath, filepath);
  }
  const xml = readFileSync(file, "utf-8");
  const json = parser.parse(xml);
  const root = json?.GSD || json?.GSDML || json;
  const metadata = {
    manufacturer: root?.Manufacturer,
    type: root?.Type,
    ident_number: root?.IdentNumber,
  };
  const io_config = parseIO(root);
  return { filepath: file, metadata, io_config };
}

function parseIO(root: any) {
  const inputs = normalize(root?.Inputs?.Input);
  const outputs = normalize(root?.Outputs?.Output);
  return {
    input_length: inputs.reduce((sum, item) => sum + Number(item?.Length || 0), 0),
    output_length: outputs.reduce((sum, item) => sum + Number(item?.Length || 0), 0),
  };
}

function normalize<T>(value: T | T[] | undefined): T[] {
  if (!value) return [];
  return Array.isArray(value) ? value : [value];
}

File: PROFIBUS-Project/profibus-python/src/profibus_mcp/gsd_parser.py
"""GSD file parser utilities."""

from __future__ import annotations

from functools import lru_cache
from pathlib import Path
from typing import Any, Dict, List, Optional

from lxml import etree


class GSDParserError(RuntimeError):
    """Raised when a GSD file cannot be parsed."""


class GSDParser:
    def __init__(self, base_path: Optional[Path | str] = None) -> None:
        self.base_path = Path(base_path).expanduser() if base_path else None

    def _resolve(self, filepath: str | Path) -> Path:
        path = Path(filepath).expanduser()
        if path.is_file():
            return path
        if self.base_path:
            candidate = self.base_path / path.name
            if candidate.is_file():
                return candidate
        raise FileNotFoundError(f"GSD file not found: {filepath}")

    def load(self, filepath: str | Path) -> Dict[str, Any]:
        path = self._resolve(filepath)
        try:
            tree = etree.parse(str(path))
        except Exception as exc:
            raise GSDParserError(f"Failed to parse GSD file {path}: {exc}") from exc
        root = tree.getroot()
        metadata = {
            "manufacturer": root.attrib.get("Manufacturer"),
            "type": root.attrib.get("Type"),
            "ident_number": root.attrib.get("IdentNumber"),
        }
        io_config = self._parse_io(root)
        return {
            "filepath": str(path),
            "metadata": metadata,
            "io_config": io_config,
        }

    def _parse_io(self, root: etree._Element) -> Dict[str, Any]:
        inputs = root.findall(".//Input")
        outputs = root.findall(".//Output")
        return {
            "input_length": sum(int(item.attrib.get("Len", 0)) for item in inputs),
            "output_length": sum(int(item.attrib.get("Len", 0)) for item in outputs),
        }

    @lru_cache(maxsize=64)
    def load_cached(self, filepath: str | Path) -> Dict[str, Any]:
        return self.load(filepath)
